gym/spaces/box.py

```diff
@@ -22,67 +24,71 @@ class Box(Space):
-                np.isscalar(low) or low.shape == shape
-            ), "low.shape doesn't match provided shape"
-            assert (
-                np.isscalar(high) or high.shape == shape
-            ), "high.shape doesn't match provided shape"
         elif not np.isscalar(low):
-            shape = low.shape
-            assert (
-                np.isscalar(high) or high.shape == shape
-            ), "high.shape doesn't match low.shape"
+            shape = low.shape  # type: ignore
         elif not np.isscalar(high):
-            shape = high.shape
-            assert (
-                np.isscalar(low) or low.shape == shape
-            ), "low.shape doesn't match high.shape"
+            shape = high.shape  # type: ignore
         else:
             raise ValueError(
                 "shape must be provided or inferred from the shapes of low or high"
             )
+        assert isinstance(shape, tuple)
 
         if np.isscalar(low):
             low = np.full(shape, low, dtype=dtype)
 
         if np.isscalar(high):
             high = np.full(shape, high, dtype=dtype)
 
-        self._shape = shape
-        self.low = low
-        self.high = high
+        assert isinstance(low, np.ndarray)
+        assert low.shape == shape, "low.shape doesn't match provided shape"
+        assert isinstance(high, np.ndarray)
+        assert high.shape == shape, "high.shape doesn't match provided shape"
+
+        self._shape: Tuple[int, ...] = shape
 
-        def _get_precision(dtype):
+        def _get_precision(dtype) -> SupportsFloat:
             if np.issubdtype(dtype, np.floating):
                 return np.finfo(dtype).precision
             else:
                 return np.inf
 
-        low_precision = _get_precision(self.low.dtype)
-        high_precision = _get_precision(self.high.dtype)
+        low_precision = _get_precision(low.dtype)
+        high_precision = _get_precision(high.dtype)
         dtype_precision = _get_precision(self.dtype)
-        if min(low_precision, high_precision) > dtype_precision:
+        if min(low_precision, high_precision) > dtype_precision:  # type: ignore
             logger.warn(f"Box bound precision lowered by casting to {self.dtype}")
-        self.low = self.low.astype(self.dtype)
-        self.high = self.high.astype(self.dtype)
+        self.low = low.astype(self.dtype)
+        self.high = high.astype(self.dtype)
 
         # Boolean arrays which indicate the interval type for each coordinate
         self.bounded_below = -np.inf < self.low
         self.bounded_above = np.inf > self.high
 
         super().__init__(self.shape, self.dtype, seed)
 
-    def is_bounded(self, manner="both"):
-        below = np.all(self.bounded_below)
-        above = np.all(self.bounded_above)
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        """Has stricter type than gym.Space - never None."""
+        return self._shape
+
+    def is_bounded(self, manner: str = "both") -> bool:
```

Comment:
I'm not a fan of the string argument here. Either an enum, or maybe some booleans instead would be nice. The current approach slightly defeats the point of adding types, since most strings don't actually make any sense as an input.

Reply:
The usability issue with an enum is that users need to import it:
if we have
```python
class Bound(Enum):
    BELOW = auto()
    ABOVE = auto()
    BOTH = auto()
```

Then either every user needs to import this Enum, or we should put it inside the Box class to be imported and available with it. We could also support both strings and Enum, but it's not a big win then. Overall the pattern of using strings in places that actually should have enums is widespread in ML, unfortunately - from losses and activations in Keras / Scikit-learn to environment creation in gym.

I would in general be in favor of revamping this part and using Enums, but it's better done in a coordinated effort and with preliminary design decision on which pattern exactly should be used.

So I suggest letting it be as it is right now, but maybe creating an issue to gather feedback on removing this pattern in gym in general. what do you think?